src/server/gc_worker/compaction_filter.rs

```diff
@@ -154,133 +216,294 @@ struct WriteCompactionFilter {
     fn new(db: Arc<DB>, safe_point: u64, context: &CompactionFilterContext) -> Self {
         // Safe point must have been initialized.
         assert!(safe_point > 0);
-        debug!(
-            "gc in compaction filter";
-            "safe_point" => safe_point, "files" => ?context.file_numbers(),
-        );
+        debug!("gc in compaction filter"; "safe_point" => safe_point);
 
-        let engine = RocksEngine::from_db(db.clone());
-        let write_batch = RocksWriteBatch::with_capacity(db, DEFAULT_DELETE_BATCH_SIZE);
-        let write_iter = {
-            // TODO: give lower bound and upper bound to the iterator.
-            let opts = IterOptions::default();
-            Some(engine.iterator_cf_opt(CF_WRITE, opts).unwrap())
-        };
-
-        WriteCompactionFilter {
+        let mut filter = WriteCompactionFilter {
             safe_point,
-            engine,
-            write_batch,
+            engine: RocksEngine::from_db(db.clone()),
+            write_batch: RocksWriteBatch::with_capacity(db, DEFAULT_DELETE_BATCH_SIZE),
+            encountered_errors: false,
+
             key_prefix: vec![],
             remove_older: false,
-            write_iter,
+
+            is_bottommost_level: context.is_bottommost_level(),
+            deleting_filtered: None,
             near_seek_distance: NEAR_SEEK_LIMIT,
+            write_iter: None,
+
             versions: 0,
             filtered: 0,
             deleted: 0,
             total_versions: 0,
             total_filtered: 0,
             total_deleted: 0,
+            versions_hist: MVCC_VERSIONS_HISTOGRAM.local(),
+            filtered_hist: GC_DELETE_VERSIONS_HISTOGRAM.local(),
+            seek_times: 0,
+            next_times: 0,
+            mvcc_delete_skip_older: 0,
+            mvcc_rollback_and_locks: 0,
+            #[cfg(test)]
+            callbacks_on_drop: {
+                let ctx = GC_CONTEXT.lock().unwrap();
+                ctx.as_ref().unwrap().callbacks_on_drop.clone()
+            },
+        };
+
+        if filter.is_bottommost_level {
+            // MVCC-delete marks can only be handled in the bottommost level.
+            // PTAL at test case `test_remove_and_skip_until`.
+            let db = filter.engine.as_inner();
+            let cf = db.cf_handle(CF_WRITE).unwrap();
+            let iter = DBIterator::new_cf(db.clone(), cf, ReadOptions::new());
+            filter.write_iter = Some(RocksEngineIterator::from_raw(iter));
         }
+        filter
     }
 
-    // Do gc before a delete mark.
-    fn handle_delete_mark(&mut self, mark: &[u8]) {
-        let mut valid = self.write_iter_seek_to(mark);
+    fn do_filter(
+        &mut self,
+        _start_level: usize,
+        key: &[u8],
+        sequence: u64,
+        value: &[u8],
+        value_type: CompactionFilterValueType,
+    ) -> Result<CompactionFilterDecision, String> {
+        self.meet_a_key_in_filter();
+
+        let (key_prefix, commit_ts) = split_ts(key)?;
```

Comment:
```suggestion
        let (user_key, commit_ts) = split_ts(key)?;
```

Reply:
All keys in this file are user keys (not RocksDB internal keys). So I think `mvcc_key_prefix` is better.